### ISSUE TITLE
Forgot password on an unconfirmed doesn't throw UserNotConfirmedException

### ISSUE THREAD

**Issue by Gidmark-wowgroup** [aws/aws-aspnet-cognito-identity-provider#179]
## Description
When running the forgot password flow on an unconfirmed user (Account Status | Enabled / UNCONFIRMED) I expect to get a UserNotConfirmedException, but at the moment everything returns "Ok", but it doesn't send a confirmation code.

## Reproduction Steps
1. Sign up a user in a Cognito user pool
2. Call ICognitoIdentityProvider's ForgotPasswordAsync with the username from step 1

### Environment
* Build Version: 3.3.110.3
* OS Info: Docker (mcr.microsoft.com/dotnet/core/aspnet:3.1-buster-slim) in Windows 10 
* Build Environment: dotnet / docker
* Targeted .NET Platform:  .NET Core 3.1

---

This is a :bug: bug-report

**Comment by Gidmark-wowgroup** [comment]
Keep open

**Comment by AzatKalim** [comment]
Also ForgotPasswordAsync doesn't throw an exception and returns Ok response if the user doesn't exist at all.

**Comment by ashishdhingra** [comment]
Upon investigating for `ForgotPasswordAsync` scenario, looks like service itself if returning HTTP OK response. This is more likely a `service-api` issue.

Used the following code which creates `UNCONFIRMED` user and then tries to execute forgot password flow:
```C#
using System;
using System.Security.Cryptography;
using System.Text;
using Amazon.CognitoIdentityProvider;
using Amazon.CognitoIdentityProvider.Model;

namespace CognitoIdentityProvider_Issue179
{
    class Program
    {
        private static string userId = "<<User_Name>>";
        private static string password = "<<User_Password>>";
        private static string email = "<<User_Email>>";
        private static string clientId = "<<appclient_Id>>";
        private static string clientSecret = "<<App_Client_secret>>";

        static void Main(string[] args)
        {
            AmazonCognitoIdentityProviderClient amazonCognitoIdentityProviderClient = new AmazonCognitoIdentityProviderClient();
            var signupResponse = amazonCognitoIdentityProviderClient.SignUpAsync(new Amazon.CognitoIdentityProvider.Model.SignUpRequest()
            {
                ClientId = clientId,
                SecretHash = GetUserPoolSecretHash("testuser2", clientId, clientSecret),
                Username = userId,
                Password = password,
                UserAttributes = new System.Collections.Generic.List<AttributeType>() { new AttributeType() { Name = "email", Value = email } }
            }).Result;

            var forgotPasswordResponse = amazonCognitoIdentityProviderClient.ForgotPasswordAsync(new Amazon.CognitoIdentityProvider.Model.ForgotPasswordRequest()
            {
                ClientId = clientId,
                SecretHash = GetUserPoolSecretHash(userId, clientId, clientSecret),
                Username = userId
            }).Result;
        }

        private static string GetUserPoolSecretHash(string userID, string clientID, string clientSecret)
        {
            string message = userID + clientID;
            byte[] keyBytes = Encoding.UTF8.GetBytes(clientSecret);
            byte[] messageBytes = Encoding.UTF8.GetBytes(message);

            HMACSHA256 hmacSha256 = new HMACSHA256(keyBytes);
            byte[] hashMessage = hmacSha256.ComputeHash(messageBytes);

            return Convert.ToBase64String(hashMessage);
        }
    }
}
```

**Comment by doc-rj-celltrak** [comment]
I think the SignUp and SignIn APIs should also throw the UserNotConfirmedException. The SignUp API currently throws UsernameExistsException if called again for an unconfirmed account... @ashishdhingra Can you confirm that this issue will cover the SignUp and SignIn APIs as well as ForgotPassword? Thanks!

**Comment by ashishdhingra** [comment]
P49047574

**Comment by ashishdhingra** [comment]
Hi @Gidmark-wowgroup,

Good evening.

I got a response from service team. The default Security configuration setting in App client prevent user existence errors from being returned from the service (refer the screenshot below): 
<img width="480" alt="Screen Shot 2021-06-29 at 7 14 36 PM" src="https://user-images.githubusercontent.com/67916761/123892030-a838db00-d90e-11eb-80ae-b703e35c5483.png">

If you change the setting to `Legacy`, service will return user existence errors to the client. I verified this by changing the setting to `Legacy` and got error `InvalidParameterException` with message `Cannot reset password for the user as there is no registered/verified email or phone_number` while trying to execute ForgotPassword flow for unconfirmed user (using the code in https://github.com/aws/aws-aspnet-cognito-identity-provider/issues/179#issuecomment-819755449). This is one of the error types returned by [ForgotPassword](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ForgotPassword.html) service API operation. Kindly note that service might return different errors based on scenario, which are beyond control of the SDK.

Hope this provides explanation on why errors were suppressed initially.

Closing this as a `guidance` issue.

Thanks,
Ashish
